### ISSUE TITLE
Value class problems

### ISSUE THREAD

**Issue by ghost** [scala/pickling#339]
I'm having a bit of trouble whenever I deal with value classes. I'm not sure if these should be classified as bugs, I'm doing something incorrectly, or if it's something not supported. Here are some of my observations with scala-pickling 0.10.1 and scala 2.11.6

I've only been using Scala Pickling a couple of days so I may be missing some understanding. I'd appreciate any clarification if possible. Thanks!
##### Ok. Both fields extend AnyVal

``` scala
import scala.pickling._
import scala.pickling.binary._
import scala.pickling.Defaults._
import scala.pickling.static._

case class A(value: Int) extends AnyVal
case class B(value: Int) extends AnyVal
case class Test(a: A, b: B)

object Example {
  def main(args: Array[String]): Unit = {
    val test = Test(A(1), B(2))
    val pickled = test.pickle
    println(pickled.unpickle[Test])
  }
}
```

```
Test(A(1),B(2))
[success] Total time: 1 s, completed 15/06/2015 3:00:29 PM
```
##### Fails when a single property doesn't extend AnyVal

``` scala
import scala.pickling._
import scala.pickling.binary._
import scala.pickling.Defaults._
import scala.pickling.static._

case class A(value: Int) extends AnyVal
case class B(value: Int)
case class Test(a: A, b: B)

object Example {
  def main(args: Array[String]): Unit = {
    val test = Test(A(1), B(2))
    val pickled = test.pickle
    println(pickled.unpickle[Test])
  }
}
```

```
[error] (run-main-72) java.lang.IllegalArgumentException: Can not set final int field Test.a to A
java.lang.IllegalArgumentException: Can not set final int field Test.a to A
        at sun.reflect.UnsafeFieldAccessorImpl.throwSetIllegalArgumentException(UnsafeFieldAccessorImpl.java:167)
        at sun.reflect.UnsafeFieldAccessorImpl.throwSetIllegalArgumentException(UnsafeFieldAccessorImpl.java:171)
        at sun.reflect.UnsafeQualifiedIntegerFieldAccessorImpl.set(UnsafeQualifiedIntegerFieldAccessorImpl.java:100)
        at java.lang.reflect.Field.set(Field.java:764)
        at Example$TestUnpickler$macro$4$2$.unpickle(Example.scala:14)
        at scala.pickling.Unpickler$class.unpickleEntry(Pickler.scala:79)
        at Example$TestUnpickler$macro$4$2$.unpickleEntry(Example.scala:14)
        at scala.pickling.functions$.unpickle(functions.scala:11)
        at scala.pickling.UnpickleOps.unpickle(Ops.scala:23)
        at Example$.main(Example.scala:14)
        at Example.main(Example.scala)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:497)
```
##### Fine when they are made final

``` scala
import scala.pickling._
import scala.pickling.binary._
import scala.pickling.Defaults._
import scala.pickling.static._

final case class A(value: Int) extends AnyVal
final case class B(value: Int)
case class Test(a: A, b: B)

object Example {
  def main(args: Array[String]): Unit = {
    val test = Test(A(1), B(2))
    val pickled = test.pickle
    println(pickled.unpickle[Test])
  }
}
```

```
Test(A(1),B(2))
[success] Total time: 1 s, completed 15/06/2015 2:57:56 PM
```
##### Breaks if one is made an Option

``` scala
import scala.pickling._
import scala.pickling.binary._
import scala.pickling.Defaults._
import scala.pickling.static._

final case class A(value: Int) extends AnyVal
final case class B(value: Int)
case class Test(a: A, b: Option[B])

object Example {
  def main(args: Array[String]): Unit = {
    val test = Test(A(1), Some(B(2)))
    val pickled = test.pickle
    println(pickled.unpickle[Test])
  }
}
```

```
[error] (run-main-82) java.lang.IllegalArgumentException: Can not set final int field Test.a to A
java.lang.IllegalArgumentException: Can not set final int field Test.a to A
        at sun.reflect.UnsafeFieldAccessorImpl.throwSetIllegalArgumentException(UnsafeFieldAccessorImpl.java:167)
        at sun.reflect.UnsafeFieldAccessorImpl.throwSetIllegalArgumentException(UnsafeFieldAccessorImpl.java:171)
        at sun.reflect.UnsafeQualifiedIntegerFieldAccessorImpl.set(UnsafeQualifiedIntegerFieldAccessorImpl.java:100)
        at java.lang.reflect.Field.set(Field.java:764)
        at Example$TestUnpickler$macro$7$2$.unpickle(Example.scala:14)
        at scala.pickling.Unpickler$class.unpickleEntry(Pickler.scala:79)
        at Example$TestUnpickler$macro$7$2$.unpickleEntry(Example.scala:14)
        at scala.pickling.functions$.unpickle(functions.scala:11)
        at scala.pickling.UnpickleOps.unpickle(Ops.scala:23)
        at Example$.main(Example.scala:14)
        at Example.main(Example.scala)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:497)
```

I'm confused at this point. So I start looking at pickling Options.

``` scala
import scala.pickling._
import scala.pickling.binary._
import scala.pickling.Defaults._
import scala.pickling.static._

final case class A(value: Int)

object Example {
  def main(args: Array[String]): Unit = {
    PicklerUnpickler.generate[Option[A]]
  }
}
```

```
[error] Example.scala:10: cannot generate fully static pickler because: 'scala.Some' allows unknown subclasses (it is no
t sealed or final isCaseClass=false isEffectivelyFinal=false isSealed=false directSubclasses=List())
[error]     PicklerUnpickler.generate[Option[A]]
```

I don't really understand why since Some appears to be final. It turns out that removing the `scala.pickling.static._` import makes it compile. 
##### Ok. I'll try to remove the static import to get Options to work with AnyVal

``` scala
import scala.pickling._
import scala.pickling.binary._
import scala.pickling.Defaults._

final case class A(value: Int) extends AnyVal
final case class B(value: Int) extends AnyVal
case class Test(a: A, b: Option[B])

object Example {
  def main(args: Array[String]): Unit = {
    val test = Test(A(1), Some(B(2)))
    val pickled = test.pickle
    println(pickled.unpickle[Test])
  }
}
```

```
[error] (run-main-8e) java.lang.IllegalArgumentException: Can not set final int field Test.a to A
java.lang.IllegalArgumentException: Can not set final int field Test.a to A
        at sun.reflect.UnsafeFieldAccessorImpl.throwSetIllegalArgumentException(UnsafeFieldAccessorImpl.java:167)
        at sun.reflect.UnsafeFieldAccessorImpl.throwSetIllegalArgumentException(UnsafeFieldAccessorImpl.java:171)
        at sun.reflect.UnsafeQualifiedIntegerFieldAccessorImpl.set(UnsafeQualifiedIntegerFieldAccessorImpl.java:100)
        at java.lang.reflect.Field.set(Field.java:764)
        at Example$TestUnpickler$macro$7$2$.unpickle(Example.scala:13)
        at scala.pickling.Unpickler$class.unpickleEntry(Pickler.scala:79)
        at Example$TestUnpickler$macro$7$2$.unpickleEntry(Example.scala:13)
        at scala.pickling.functions$.unpickle(functions.scala:11)
        at scala.pickling.UnpickleOps.unpickle(Ops.scala:23)
        at Example$.main(Example.scala:13)
        at Example.main(Example.scala)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:497)
```
##### It only works without a property extending AnyVal

``` scala
import scala.pickling._
import scala.pickling.binary._
import scala.pickling.Defaults._

final case class A(value: Int)
final case class B(value: Int)
case class Test(a: A, b: Option[B])

object Example {
  def main(args: Array[String]): Unit = {
    val test = Test(A(1), Some(B(2)))
    val pickled = test.pickle
    println(pickled.unpickle[Test])
  }
}
```

```
Test(A(1),Some(B(2)))
[success] Total time: 1 s, completed 15/06/2015 3:30:29 PM
```

**Comment by jsuereth** [comment]
This is actually a bug in scala-reflection I just discovered.   The code you list works in 0.11.x branch now since we use the new `scala.pickling.internal.Reflect` method to find private members.

Basically previously, scala-reflection _fails_ to correctly discover private val/vars.  I'm not sure how you hit this bug, but I just ran a test against master (with the new macros/reflection) and it all appears to be solved.

Thanks for the report!  cc> @phaller
